### PR TITLE
sync node topology's label to vmi

### PR DESCRIPTION
**Pull request by chrisho**
**Problem:**
Sync node  topology's label to VMI

**Solution:**
When the VMI is created, the node topology label will be synced to the VMI annotation

**Related Issue:**
#1418 

**Test plan:**
https://github.com/harvester/cloud-provider-harvester/pull/21

### CODE REVIEW

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 1]
```diff
@@ -12,10 +15,21 @@ import (
 	kubevirtctrl "github.com/harvester/harvester/pkg/generated/controllers/kubevirt.io/v1"
 )
 
+const (
+	VirtualMachineCreatorHarvester = "docker-machine-driver-harvester"
```
Comment (futuretea) [inline pkg/controller/master/virtualmachine/vmi_controller.go]: VirtualMachineCreatorNodeDriver = "docker-machine-driver-harvester"

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 2]
```diff
@@ -2,8 +2,11 @@ package virtualmachine
 
 import (
 	"fmt"
+	"reflect"
 
-	v1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/harvester/harvester/pkg/builder"
```
Comment (futuretea) [inline pkg/controller/master/virtualmachine/vmi_controller.go]: packages from this repo should be put last

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 3]
```diff
@@ -91,3 +105,38 @@ func (h *VMIController) UnsetOwnerOfPVCs(_ string, vmi *kubevirtv1.VirtualMachin
...
 	return vmi, nil
 }
+
+// ReconcileFromHostLabels handles the propagation of metadata from node labels to VirtualMachineInstance annotations.
+func (h *VMIController) ReconcileFromHostLabels(_ string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmi == nil || vmi.DeletionTimestamp != nil {
+		return vmi, nil
+	}
+
+	if v, exist := vmi.Labels[builder.LabelKeyVirtualMachineCreator]; !exist || v != VirtualMachineCreatorHarvester {
```
Comment (futuretea) [inline pkg/controller/master/virtualmachine/vmi_controller.go]:
```go
	if creator := vmi.Labels[builder.LabelKeyVirtualMachineCreator]; creator != VirtualMachineCreatorNodeDriver {
		return vmi, nil
	}
```

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 4]
```diff
@@ -91,3 +105,38 @@ func (h *VMIController) UnsetOwnerOfPVCs(_ string, vmi *kubevirtv1.VirtualMachin
...
+		return vmi, nil
+	}
+
+	node, err := h.nodeCache.Get(vmi.Status.NodeName)
+	if err != nil && !apierrors.IsNotFound(err) {
```
Comment (futuretea) [inline pkg/controller/master/virtualmachine/vmi_controller.go]:
```go
	if  err != nil {
		if apierrors.IsNotFound(err) {
			return vmi, nil
		}
		return vmi, fmt.Errorf("failed to get node %s for VirtualMachineInstance %s: %v", vmi.Status.NodeName, vmi.Name, err)
	}
```

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 5]
```diff
@@ -91,3 +105,38 @@ func (h *VMIController) UnsetOwnerOfPVCs(_ string, vmi *kubevirtv1.VirtualMachin
...
+		return vmi, fmt.Errorf("failed to get node %s for VirtualMachineInstance %s: %v", vmi.Status.NodeName, vmi.Name, err)
+	} else if apierrors.IsNotFound(err) {
+		return vmi, nil
+	}
+
+	toUpdate := vmi.DeepCopy()
+	for _, label := range hostLabelsReconcileMapping {
+		srcValue, srcExists := node.Labels[label]
+		if _, exist := toUpdate.Annotations[label]; !srcExists && exist {
```
Comment (futuretea) [inline pkg/controller/master/virtualmachine/vmi_controller.go]: if !srcExists && !exist ,we should not do any thing

```go
		if srcExists {
			toUpdate.Annotations[label] = srcValue
		} else if _, exist := toUpdate.Annotations[label]; exist {
			delete(toUpdate.Annotations, label)
		}
```

File: pkg/controller/master/virtualmachine/vmi_controller.go [hunk 6]
```diff
@@ -91,3 +105,38 @@ func (h *VMIController) UnsetOwnerOfPVCs(_ string, vmi *kubevirtv1.VirtualMachin
...
+			return vmi, nil
+		}
+		return vmi, fmt.Errorf("failed to get node %s for VirtualMachineInstance %s: %v", vmi.Status.NodeName, vmi.Name, err)
+	}
+
+	toUpdate := vmi.DeepCopy()
+	for _, label := range hostLabelsReconcileMapping {
+		srcValue, srcExists := node.Labels[label]
+		if srcExists {
+			toUpdate.Annotations[label] = srcValue
```
Comment (futuretea) [inline pkg/controller/master/virtualmachine/vmi_controller.go:134]: need a nil check before write map
```go
			if toUpdate.Annotations == nil {
				toUpdate.Annotations = map[string]string{}
			}
```

**Review by futuretea — APPROVED** [review]
LGTM

**Review by guangbochen — APPROVED** [review]
LGTM. Needs to backport to v1.0.3 and add related docs of this feature enhancement, thanks.

**Pull request merged**